### PR TITLE
tests: has: Added support for testing Hearing Access Service.

### PR DESCRIPTION
**Pull request by MSEEHenrik**
Currently no support for synchronized presets.

### CODE REVIEW

File: doc/btp_has.txt
Comment (sjanc) [inline doc/btp_has.txt:83]: clarify if string should be null terminated or not (I guess null is not needed since we have length param)

File: doc/btp_has.txt
Comment (sjanc) [inline doc/btp_has.txt:27]: nitpick,  adjust indentation,  in other doc files we typically use tabs (8 spaces wide)

File: autopts/wid/has.py
Comment (sjanc) [inline autopts/wid/has.py:26]: this should probably be in autopts/pybtp/defs.py  (and I'd prefer it to be 0x00, 0x01, 0x02 etc)

**Review by sjanc — COMMENTED** [review]
Hi,

this looks really good!  some minor comment though in code

And please rebase your branch (ie no merge commits in PR)

File: autopts/wid/has.py
Comment (MSEEHenrik) [reply autopts/wid/has.py]: Done, but kept it locally in autopts/wid/has.py, since it is only used here.  Prefer no to "pollute" autopts/pybtp/defs.py.

**Comment by mkasenberg** [comment]
Maybe the HAS tests could be enabled in the PTS zephyr-master workspace? Then they will be included in the zephyr weekly test run.

File: autopts/wid/has.py
Comment (mkasenberg) [inline autopts/wid/has.py:42]: Lets use recently introduced generic wid that fixed some misleading exception:
```suggestion
def has_wid_hdl(wid, description, test_case_name):
    log(f'{has_wid_hdl.__name__}, {wid}, {description}, {test_case_name}')
    return generic_wid_hdl(wid, description, test_case_name, [__name__])
```

File: autopts/wid/has.py
Comment (mkasenberg) [inline autopts/wid/has.py:54]: Pycharm IDE highlights me some warnings about Python PEP 8 standard, which we try to follow (most of the time ;) ):
- comments should be inside the functions
- put 2 blank lines between functions
- put a space after a colon `_: WIDParams`
```suggestion
def hdl_wid_450(_: WIDParams):
    """
    Please add a new Preset Record.
    """
    n, properties = 4, PresetProperty.BT_HAS_PROP_AVAILABLE | PresetProperty.BT_HAS_PROP_WRITABLE
    btp.has_add_preset(n, properties, 'PRESET_' + str(n))
    return True


```

File: autopts/wid/has.py
Comment (MSEEHenrik) [reply autopts/wid/has.py]: Done.

File: autopts/ptsprojects/zephyr/has_wid.py
Comment (MSEEHenrik) [reply autopts/ptsprojects/zephyr/has_wid.py]: Done.

File: autopts/wid/has.py
Comment (MSEEHenrik) [reply autopts/wid/has.py]: Done.

**Comment by MSEEHenrik** [comment]
> Maybe the HAS tests could be enabled in the PTS zephyr-master workspace? Then they will be included in the zephyr weekly test run.

Yes - absolutely.  I will update zephyr-master as soon as the two PRs have been merged.

**Comment by mkasenberg** [comment]
I tested with default ICSes/IXITs for HAS workspace, so they may need adjustment, but nice score:
<details><summary>Failed tests</summary>
HAS     HAS/SR/FEAT/BV-04-C             FAIL<br>
HAS     HAS/SR/FEAT/BV-06-C             FAIL<br>
HAS     HAS/SR/FEAT/BV-08-C             FAIL<br>
HAS     HAS/SR/SGGIT/CHA/BV-04-C        FAIL<br>
</details>
<details><summary>Successful tests</summary>
HAS     HAS/SR/CFG/BV-01-C              PASS <br>
HAS     HAS/SR/CP/BV-01-C               PASS<br>
HAS     HAS/SR/CP/BV-05-C               PASS<br>
HAS     HAS/SR/CP/BV-07-C               PASS<br>
HAS     HAS/SR/CP/BV-09-C               PASS<br>
HAS     HAS/SR/CP/BV-11-C               PASS<br>
HAS     HAS/SR/FEAT/BV-02-C             PASS<br>
HAS     HAS/SR/SGGIT/SER/BV-01-C        PASS<br>
HAS     HAS/SR/SPE/BI-01-C              PASS<br>
HAS     HAS/SR/SPE/BI-02-C              PASS<br>
HAS     HAS/SR/SPE/BI-07-C              PASS<br>
HAS     HAS/SR/SPE/BI-10-C              PASS<br>
HAS     HAS/SR/SPE/BI-13-C              PASS<br>
HAS     HAS/SR/SPE/BI-14-C              PASS<br>
</details>

**Review by mkasenberg — APPROVED** [review]

**Comment by MSEEHenrik** [comment]
> 

mkasenberg - thanks for the effort.  I will look into the failing tests.  I only saw two errors during my run, but perhaps our ICS's are a bit different.

**Comment by MariuszSkamra** [comment]
Related Zephyr PR: https://github.com/zephyrproject-rtos/zephyr/pull/58304
Let's wait till it's merged

**Review by MariuszSkamra — APPROVED** [review]

**Pull request merged**
